src/qmapshack/device/CDeviceWatcherLinux.cpp

```diff
@@ -157,30 +156,54 @@ QString CDeviceWatcherLinux::readMountPoint(const QString& path) {
     message.setArguments(args);
   }
 
-  QDBusMessage reply = QDBusConnection::systemBus().call(message);
+  const QDBusMessage& reply = QDBusConnection::systemBus().call(message);
 
 #if defined(Q_OS_FREEBSD)
+  // this is probably broken code and needs to be fixed
   for (const QVariant& arg : reply.arguments()) {
     if (!arg.value<QDBusVariant>().variant().value<QStringList>().isEmpty()) {
       points.append(arg.value<QDBusVariant>().variant().value<QStringList>().first());
     }
   }
 #else
-  QList<QByteArray> list;
-  {
-    const QList<QVariant>& args = reply.arguments();
-    for (const QVariant& arg : args) {
-      arg.value<QDBusVariant>().variant().value<QDBusArgument>() >> list;
+  QStringList mountPoints;
+
+  for (const QVariant& arg : reply.arguments()) {
+    // This seems to be the only way to get the underlying DBus data type.
+    // It's is a very poor design. Maybe we do not understand Qt.
+    const QVariant& dbusVar = arg.value<QDBusVariant>().variant();
+    const QDBusArgument& dbusArg = dbusVar.value<QDBusArgument>();
+
+    // Test for the correct signature. It's not really expected to fail.
+    // But better safe than sorry
+    if (dbusArg.currentSignature() != "aay") {
+      continue;
     }
-  }
 
-  for (const QByteArray& point : std::as_const(list)) {
-    points.append(point);
+    // As the signature is "aay" it's an array of an array of bytes.
+    // This translates to an array of QByteArray. Don't know why there
+    // should be several mount points and if that is ever happening.
+    // But we do not expect less wackiness from a DBus definition.
+
+    // That is the way Qt wants us to iterate
+    dbusArg.beginArray();
+    while (!dbusArg.atEnd()) {
+      QByteArray data;
+      dbusArg >> data;
+      // Not sure about this one. But my guts feeling tells me that
+      // we should decode the stuff according to the system's requirements.
+      auto decoder = QStringDecoder(QStringDecoder::System);
+      // As QByteArray has a trailing 0 however QString does not it needs to be
+      // removed. Else the QString has a 0 as unicode character, confusing the
+      // rest of the system.
```

Comment:
This sounds like the `0` is added by QByteArray, but it's the last entry in the array coming from the DBus.

Reply:
According to the docs QByteArray always enforces a trailing 0 . If there is one in the raw data (DBus message) it's ok. If not it is appended by QByteArray. So it'S always guaranteed to be 0 terminated. 

The part I do not understand is why QString does not recognize it as 0 terminated string and takes over the 0 as part of the string.

Reply:
From my understanding QByteArray represents just a number of pure bytes, whereas QString is always UTF16 encoded. From [Using QByteArray](https://wiki.qt.io/Using_QByteArray):

> To convert a QByteArray to QString, one needs to specify the encoding. It is not possible to convert a QByteArray to QString without knowing the encoding. Once the encoding is known, one can use functions like QString::fromUtf8, QString::fromUcs4 or use QTextCodec::toUnicode().

But i have no experience in QT.

Reply:
Yes decoding is another caveat when converting from QByteArray to QString. That is why I introduced the decoder. Anyways this does not explain why the trailing 0 is converted to an explicit "\0" as part of the string.

Reply:
You should ask this in a QT forum…

In python it's the same, i used the int's from my qdbus output:
~~~~
>>> message = [47, 114, 117, 110, 47, 109, 101, 100, 105, 97, 47, 107, 97, 112, 117, 116, 116, 110, 105, 107, 47, 75, 65, 82, 84, 69, 32, 54, 48, 71, 66, 0]
>>> ar = bytearray(message)
>>> print(ar)
bytearray(b'/run/media/kaputtnik/KARTE 60GB\x00')
>>> s = str(ar)
>>> print(s)
bytearray(b'/run/media/kaputtnik/KARTE 60GB\x00')
>>> print(ar.decode())
/run/media/kaputtnik/KARTE 60GB
>>> s = str(ar, encoding='utf-8')
>>> print(s)
/run/media/kaputtnik/KARTE 60GB
~~~~

Reply:
I think the correct way to access the byte array is to use the `const char*` pointer. By that decoding can detect the null termination properly.